File: src/Deck.ts
import { Card, Suit, Rank } from "./Card";

export class Deck {
  private cards: Card[] = [];

  constructor() {
    this.reset();
  }

  reset(): void {
    this.cards = [];
    const suits = Object.values(Suit) as Suit[];
    const ranks = Object.values(Rank).filter(v => typeof v === 'number') as Rank[];

    for (const suit of suits) {
      for (const rank of ranks) {
        this.cards.push(new Card(suit, rank));
      }
    }
    this.shuffle();
  }

  shuffle(): void {
    for (let i = this.cards.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]];
    }
  }

  draw(): Card {
    if (this.cards.length === 0) {
      this.reset();
    }
    return this.cards.pop()!;
  }

  getCardsLeft(): number {
    return this.cards.length;
  }
}

File: Assets/Scripts/Deck.cs
using System.Collections.Generic;
using UnityEngine;

public class Deck
{
    public List<CardData> Cards { get; private set; }

    public Deck()
    {
        Reset();
    }

    public void Reset()
    {
        Cards = new List<CardData>();

        foreach (Suit suit in System.Enum.GetValues(typeof(Suit)))
        {
            foreach (Rank rank in System.Enum.GetValues(typeof(Rank)))
            {
                Cards.Add(new CardData { Suit = suit, Rank = rank });
            }
        }

        Shuffle();
    }

    public void Shuffle()
    {
        for (int i = 0; i < Cards.Count; i++)
        {
            int j = Random.Range(i, Cards.Count);
            var tmp = Cards[i];
            Cards[i] = Cards[j];
            Cards[j] = tmp;
        }
    }

    public CardData Draw()
    {
        if (Cards.Count == 0)
        {
            Reset();
        }

        var card = Cards[Cards.Count - 1];
        Cards.RemoveAt(Cards.Count - 1);
        return card;
    }
}
